### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by internettrans**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,215 @@
...
+import { createStore, applyMiddleware } from 'redux';
+import phunk from 'redux-phunk';
+
+const store = createStore(reducer, applyMiddleware(phunk));
+
+store.dispatch(myAction());
+```
+
+## Phunks?? That's not even a word!
+A "phunk" is just an object with `name` and `phunk` properties. The `phunk` property is a function that returns an asynchronous value, such as a Promise or a (single-valued) Observable. Examples:
```
Comment (blittle) [inline README.md]: Maybe add a little more clarity with the differences between a `phunk` and an `action`. Do `phunks` get dispatched?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,215 @@
...
+      name: 'userPreferences',
+      async phunk(valueOf) {
+        const userPreferencesId = valueOf('user').preferencesId;
+        return await fetch(`/api/user-preferences/${userPreferencesId}`)
+          .then(response => response.json())
+      }
+    },
+    {
+      type: 'NEW_USER_PREFERENCES',
+      preferences: asyncValue('userPreferences'),
```
Comment (blittle) [inline README.md:149]: Haven't yet looked deep into the code, but what if two separate `phunk` sequences are fired at the same time, does `asyncValue` allow you to pull values from another sequence?

Comment (internettrans) [reply README.md:149]: If two separate phunk sequences are fired at the same time, `asyncValue` does _not_ allow you to pull values from a different sequence.

See [what asyncValue actually does](https://github.com/CanopyTax/redux-phunk/blob/start/src/redux-phunk.js#L21), where async values are given a [concrete value](https://github.com/CanopyTax/redux-phunk/blob/start/src/redux-phunk.js#L88), and where the concrete value comes from ([see](https://github.com/CanopyTax/redux-phunk/blob/start/src/redux-phunk.js#L93) and [here](https://github.com/CanopyTax/redux-phunk/blob/start/src/redux-phunk.js#L6)).

File: README.md (hunk 1 above)
Comment (internettrans) [reply README.md]: Good idea, I'll update the readme. `Phunks` do _not_ get dispatched - only actions do.

File: README.md (hunk 2 above)
Comment (blittle) [reply README.md:149]: Good, I don't think they should bleed values between sequences.

File: src/redux-phunk.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
+import { flattenDeep } from 'lodash';
+
+const phunkMiddleware = store => next => action => {
+	if (Array.isArray(action)) {
+		const phunkuence = flattenDeep(action);
+		phunkuence.phunkValues = {};
+		return new Promise((resolve, reject) => {
+			executePhunkuence(phunkuence, store.dispatch, next, resolve, reject);
```
Comment (blittle) [inline src/redux-phunk.js:8]: `phunkuence`?

**Comment by blittle** [comment]
Would it be worthwhile to split this into multiple files?

File: src/redux-phunk.js (hunk 3 above)
Comment (internettrans) [reply src/redux-phunk.js:8]: Yes. Phunkuence.

I don't understand your question.

**Comment by internettrans** [comment]
I put this together thinking that I'd start with one monolithic file and the split out into modules if I felt like it was getting out of hand. At ~150 lines, it's pretty close to splitting up but I was still wasn't feeling too bad about it. Where I am with the project is that I want to see how useful it is and how much more complicated it gets before getting serious about setting up a directory and file structure for it. Also I think what I was trying to avoid is the Michael Jackson / Ryan Florence way of writing code, where each function is its own file.

Anyway, that's a long-winded way of saying let's maybe split it up into multiple files in subsequent pull requests if it keeps on getting bigger and more complicated.

**Comment by blittle** [comment]
:camel:

**Pull request merged**